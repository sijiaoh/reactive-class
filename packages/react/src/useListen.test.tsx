import {create, act, ReactTestRenderer} from 'react-test-renderer';
import {useListen, ReactiveClass} from '.';

class Example extends ReactiveClass {
  num = 0;
  arr: string[] = [];
  child?: Example;
  children: Example[] = [];
}

describe(useListen.name, () => {
  it('rerender component when instance value changed', () => {
    let renderCount = 0;
    let expectRenderCount = 0;
    const Component = ({e, name}: {e: Example; name: string}) => {
      const eData = useListen(e);
      renderCount++;
      return (
        <>
          {name}
          {eData.num}
          {JSON.stringify(eData.arr)}
          {e.child && (
            <Component
              e={e.child}
              name={name === 'e' ? 'child' : 'grandChild'}
            />
          )}
          {e.children.map((child, index) => (
            <Component key={index} e={child} name={`children-${index}`} />
          ))}
        </>
      );
    };

    const e = new Example();
    e.child = new Example();
    e.children = [new Example(), new Example()];
    e.child.child = new Example();

    const instanceNum = 5;

    let testRenderer!: ReactTestRenderer;
    void act(() => {
      testRenderer = create(<Component e={e} name="e" />);
    });
    expect(testRenderer.toJSON()).toMatchSnapshot('0 default');
    expectRenderCount += instanceNum;
    expect(renderCount).toBe(expectRenderCount);

    void act(() => {
      e.num++;
    });
    expect(testRenderer.toJSON()).toMatchSnapshot('1 e num increment');
    expectRenderCount += instanceNum;
    expect(renderCount).toBe(expectRenderCount);

    void act(() => {
      e.arr.push('no');
    });
    expect(testRenderer.toJSON()).toMatchSnapshot('2 no changes');
    expect(renderCount).toBe(expectRenderCount);

    void act(() => {
      e.arr = ['yes'];
    });
    expect(testRenderer.toJSON()).toMatchSnapshot('3 e array changed');
    expectRenderCount += instanceNum;
    expect(renderCount).toBe(expectRenderCount);

    void act(() => {
      e.child!.num++;
    });
    expect(testRenderer.toJSON()).toMatchSnapshot('4 child num increment');
    expectRenderCount += instanceNum;
    expect(renderCount).toBe(expectRenderCount);

    void act(() => {
      e.children[0].num++;
    });
    expect(testRenderer.toJSON()).toMatchSnapshot('5 children-0 num increment');
    expectRenderCount++;
    expect(renderCount).toBe(expectRenderCount);

    void act(() => {
      e.child!.child!.num++;
    });
    expect(testRenderer.toJSON()).toMatchSnapshot('6 grandChild num increment');
    expectRenderCount += instanceNum;
    expect(renderCount).toBe(expectRenderCount);
  });
});
