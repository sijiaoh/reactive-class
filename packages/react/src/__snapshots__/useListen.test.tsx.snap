// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useListen with selector rerender component only selected value changed: 0 default 1`] = `"[]"`;

exports[`useListen with selector rerender component only selected value changed: 1 arr changed 1`] = `"[\\"yes\\"]"`;

exports[`useListen without selector rerender component when instance value changed: 0 default 1`] = `
Array [
  "e",
  "0",
  "[]",
  "grandChild: 0",
  "child",
  "0",
  "[]",
  "grandChild",
  "0",
  "[]",
  "children-0",
  "0",
  "[]",
  "children-1",
  "0",
  "[]",
]
`;

exports[`useListen without selector rerender component when instance value changed: 1 e num increment 1`] = `
Array [
  "e",
  "1",
  "[]",
  "grandChild: 0",
  "child",
  "0",
  "[]",
  "grandChild",
  "0",
  "[]",
  "children-0",
  "0",
  "[]",
  "children-1",
  "0",
  "[]",
]
`;

exports[`useListen without selector rerender component when instance value changed: 2 no changes 1`] = `
Array [
  "e",
  "1",
  "[]",
  "grandChild: 0",
  "child",
  "0",
  "[]",
  "grandChild",
  "0",
  "[]",
  "children-0",
  "0",
  "[]",
  "children-1",
  "0",
  "[]",
]
`;

exports[`useListen without selector rerender component when instance value changed: 3 e array changed 1`] = `
Array [
  "e",
  "1",
  "[\\"yes\\"]",
  "grandChild: 0",
  "child",
  "0",
  "[]",
  "grandChild",
  "0",
  "[]",
  "children-0",
  "0",
  "[]",
  "children-1",
  "0",
  "[]",
]
`;

exports[`useListen without selector rerender component when instance value changed: 4 child num increment 1`] = `
Array [
  "e",
  "1",
  "[\\"yes\\"]",
  "grandChild: 0",
  "child",
  "1",
  "[]",
  "grandChild",
  "0",
  "[]",
  "children-0",
  "0",
  "[]",
  "children-1",
  "0",
  "[]",
]
`;

exports[`useListen without selector rerender component when instance value changed: 5 children-0 num increment 1`] = `
Array [
  "e",
  "1",
  "[\\"yes\\"]",
  "grandChild: 0",
  "child",
  "1",
  "[]",
  "grandChild",
  "0",
  "[]",
  "children-0",
  "1",
  "[]",
  "children-1",
  "0",
  "[]",
]
`;

exports[`useListen without selector rerender component when instance value changed: 6 grandChild num increment 1`] = `
Array [
  "e",
  "1",
  "[\\"yes\\"]",
  "grandChild: 1",
  "child",
  "1",
  "[]",
  "grandChild",
  "1",
  "[]",
  "children-0",
  "1",
  "[]",
  "children-1",
  "0",
  "[]",
]
`;
